chrome/android/java/src/org/chromium/chrome/browser/adblock/AdblockInitializer.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * Orca is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Orca.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.chromium.chrome.browser.adblock;
+
+import android.content.Context;
+import android.support.annotation.UiThread;
+
+import org.adblockplus.libadblockplus.android.AdblockEngine;
+import org.adblockplus.libadblockplus.android.SingleInstanceEngineProvider;
+import org.adblockplus.libadblockplus.android.settings.AdblockHelper;
+import org.chromium.base.ContextUtils;
+import org.chromium.base.ThreadUtils;
+import org.chromium.base.library_loader.ProcessInitException;
+import org.chromium.chrome.browser.init.BrowserParts;
+import org.chromium.chrome.browser.init.ChromeBrowserInitializer;
+import org.chromium.chrome.browser.init.EmptyBrowserParts;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public final class AdblockInitializer implements SingleInstanceEngineProvider.EngineCreatedListener {
+
+    private final InitializationState mUnitializedState = new UnitializedState();
+    private final InitializationState mInitializedState = new InitializedState();
+    private InitializationState mInitializationState = mUnitializedState;
+    private final Set<OnAdblockInitListener> mInitListeners = new HashSet<>();
+
+    private AdblockInitializer() {
+    }
+
+    // Thread-safe lazy initialization
+    private static class LazyLoader {
+        private static final AdblockInitializer sInstance = new AdblockInitializer();
+    }
+
+    public static AdblockInitializer getInstance() {
+        return AdblockInitializer.LazyLoader.sInstance;
+    }
+
+    @Override
+    public void onAdblockEngineCreated(final AdblockEngine engine) {
+        ThreadUtils.postOnUiThread(() -> beInitialized(engine));
+    }
+
+    @UiThread
+    public void init(final ChromeBrowserInitializer initializer) {
+        mInitializationState.init(initializer);
+    }
+
+    public void addAdblockInitListener(final OnAdblockInitListener listener) {
+        mInitializationState.addAdblockInitListener(listener);
+    }
+
+    private void initAdblock() {
+        final Context context = ContextUtils.getApplicationContext();
+        final String basePath = context.getDir(
+                AdblockEngine.BASE_PATH_DIRECTORY,
+                Context.MODE_PRIVATE)
+                .getAbsolutePath();
+
+        final long ptr = AdblockBridge.getInstance().getIsolateProviderNativePtr();
+        AdblockHelper.get()
+                .init(context, basePath, true, AdblockHelper.PREFERENCE_NAME)
+                .addEngineCreatedListener(this)
+                .useV8IsolateProvider(ptr);
+
+        AdblockHelper.get().getProvider().retain(true);
```

Comment:
`retain()` increases instances count and actually creates engine if counter = 1.
The idea is that every client increases and decreases counter by calling `retain/release`. So for example when Preferences are shown, counter=1 as 2 clients are registered: `ChromeTabbedActivity` and `Preferences` activity.

Where is `release()` called here?

Reply:
I initially thought about retaining() / releasing() the AdblockEngine, The thing is, there are several flows which could lead into some of the activities in the Chromium project (as one example, there is a feature called custom tabs, which you can set the browser as default, and open a link from other apps as a standalone tab). Probably, whenever the app is being used we want the Abdlock Engine to be working. For that, in order to have edge cases where the AdblockEngine is not initialized, or even some crashes because of that, I think is better that we keep the AdblockEngine as an Application level initialization, not as a Activity level.

Reply:
Application does not have `release` event as we all know and we can't keep engine in memory as it takes too much. So retail/release was introduced intentionally to manage this precious resource. I think we need 'release' somewhere here.